Fix OpenGL segmentation fault on Linux

jclarkeSTFC opened the pull request:
When drawing some instruments in full 3d (with OpenGL), putting all the draw commands for the instrument in one big list seems to cause a segmentation fault on certain versions of Linux, including the version used on IDAaaS. This changes `InstrumentRenderer` to draw each component of the instrument using a separate list, when the `MantidOptions.InstrumentView.MesaBugWorkaround` option in the user properties file is set to `On`.

The original bug in Mesa was fixed [here](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/27556), but that version has not made it onto IDAaaS yet.

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
- Draw each instrument component in a separate display list when the new option is enabled.
- If option is disabled then behaviour should be the same.
- Add a method that will give you the OpenGL version string (someone will thank me in the future).
- Split the picking and non-picking display lists into named member variables instead of using two indices.

Fixes #36220 

### To test:
See #36220, or what caused the original segmentation fault for me was to view full 3D for MAR11001 (on Rocky 8 IDAaaS and Ubuntu). Other instruments and files will need checking as well.

Updated conda packages have bee uploaded to thomashampson/label/mesa_workaround:
`mamba install -c thomashampson/label/opengl_segfault_fix mantidworkbench`
Install this on IDAaaS and check that it doesn't crash.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
New method to report OpenGL version

Not used but useful for debugging and will save time in the future if
anyone needs to work out how to do this.

jclarkeSTFC committed:
Draw each component in separate display list

When drawing some instruments in full 3d (with OpenGL), putting all the
draw commands for the instrument in one big list seems to cause a
segmentation fault on certain versions of Linux, including the version
used on IDAaaS. This changes InstrumentRenderer to draw each component
of the instrument using a separate list.

jclarkeSTFC committed:
Add release notes

jclarkeSTFC committed:
Remove code duplication when clearing lists

jhaigh0 reviewed (changes requested):
One small thing.
I find it not so easy to ready, but I reckon that's mostly just because its complex in nature. I wonder if some comments explaining how it works would be helpful? I'll let other reviewers chime in, I don't mind either way.

RichardWaiteSTFC commented:
Tested this on MARI > Excitations Single Crystal (no GPU) and found no crash (started from terminal with `mantidworkbench`)

jclarkeSTFC committed:
Put new drawing method behind config option

Keep the old drawing method in the InstrumentRendererClassic class,
while the new method is in InstrumentRendererMultiList. Then we can
switch between methods depending on a setting in the config file.

jclarkeSTFC committed:
Update cppcheck suppression

jclarkeSTFC committed:
Add new property description to docs

jhaigh0 commented on qt/widgets/instrumentview/src/InstrumentRendererClassic.cpp:
```diff
@@ -0,0 +1,74 @@
+  const auto &compInfo = instrActor().componentInfo();
+  std::vector<bool> visited(compInfo.size(), false);
+
+  for (size_t i = compInfo.root(); i != std::numeric_limits<size_t>::max(); --i) {
+    drawComponent(i, visibleComps, showGuides, picking, compInfo, visited);
+  }
```
I know this has been copied from the old code so don't worry with fixing it now, but there's got to be a better way to do this 🙃 I think the component info class should have a way of retrieving the max index

jhaigh0 commented on qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp:
```diff
@@ -0,0 +1,89 @@
+  m_useNonPickingDisplayList = false;
+  m_usePickingDisplayList = false;
+
+  m_pickingDisplayListId.resize(componentInfo.size());
+  m_nonPickingDisplayListId.resize(componentInfo.size());
+  for (size_t i = 0; i < componentInfo.size(); ++i) {
+    m_nonPickingDisplayListId[i] = 0;
+    m_pickingDisplayListId[i] = 0;
+  }
```
There might be a simpler way to initialise these vectors
```
std::vector<T> myVector(length, repeatingElementOfTypeT);
```

jhaigh0 commented on qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp:
```diff
@@ -0,0 +1,89 @@
+
+void InstrumentRendererMultiList::renderInstrument(const std::vector<bool> &visibleComps, bool showGuides,
+                                                   bool picking) {
+  if (std::none_of(visibleComps.cbegin(), visibleComps.cend(), [](bool visible) { return visible; }))
+    return;
```
```suggestion
  if (std::none_of(visibleComps.cbegin(), visibleComps.cend()))
    return;
```

jhaigh0 commented on qt/widgets/instrumentview/src/InstrumentRendererMultiList.cpp:
```diff
@@ -0,0 +1,89 @@
+
+  OpenGLError::check("InstrumentActor::draw()");
+  const auto &displayListIds = picking ? m_pickingDisplayListId : m_nonPickingDisplayListId;
+  auto &useDisplayList = picking ? m_usePickingDisplayList : m_useNonPickingDisplayList;
```
Feels as if maybe these bools could be simplified? is it always either `m_usePickingDisplayList` or `m_useNonPickingDisplayList`? In that case could they be one bool? I might be misunderstanding.

jhaigh0 reviewed (changes requested):
A few minor suggestions, mostly to do with the older code, but looking pretty much there

jclarkeSTFC committed:
Apply suggestions from code review

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

jhaigh0 reviewed (approved):
We've had a test in IDAaaS now that they've added the config option to the local properties. Prevents the crash as intended.
